integrations/kube-agent-updater/pkg/controller/statefulset.go

```diff
@@ -0,0 +1,284 @@
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		Complete(r)
+}
+
+func (r *StatefulSetVersionUpdater) unblockStatefulSetRollout(ctx context.Context, sts *appsv1.StatefulSet) error {
+	log := ctrllog.FromContext(ctx)
+	verboseLog := ctrllog.FromContext(ctx).V(1)
+
+	if sts.Status.UpdateRevision == "" {
+		return trace.BadParameter("statefulset '%s' UpdateRevision empty", sts.Name)
+	}
+
+	// First we get all pods managed by the StatefulSet
+	stsLabelSelector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var managedPodsList v1.PodList
+	listOptions := []kclient.ListOption{kclient.InNamespace(sts.Namespace), kclient.MatchingLabelsSelector{Selector: stsLabelSelector}}
+	err = r.List(ctx, &managedPodsList, listOptions...)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// *v1.Pod implements metav1.Object, it's easier to convert now from
+	// []v1.Pod to []*v1.Pod. e.g. this is required for listNames()
+	managedPods := podListToListOfPods(&managedPodsList)
+
+	verboseLog.Info("statefulset managed pods", "managedPodsList", listNames(managedPods))
+
+	// We filter out malformed pods, not created by the statefulset
+	// We don't want to destroy other pods in case of label conflicts
+	var saneManagedPods []*v1.Pod
+	for _, pod := range managedPods {
+		if (pod.Labels == nil) || (pod.Labels[appsv1.ControllerRevisionHashLabelKey] == "") {
+			log.Error(
+				trace.Errorf("pod selected by the statefulset '%s' but does not have the '%s' label", sts.Name, appsv1.ControllerRevisionHashLabelKey),
+				"ignoring malformed pod", "podName", pod.Name, "podLabels", pod.Labels,
+			)
+		} else {
+			saneManagedPods = append(saneManagedPods, pod)
+		}
+	}
+
+	// Then we filter out all the healthy pods
+	unhealthyPods := UnhealthyPods(saneManagedPods)
+
+	if len(unhealthyPods) == 0 {
+		verboseLog.Info("no statefulset unhealthy pods")
+		return nil
+	}
+	verboseLog.Info("statefulset unhealthy pods", "unhealthyPods", listNames(unhealthyPods))
+
+	// We filter out pods issued form the latest controller revision.
+	// If those are unhealthy there's nothing we can do but hope the next update
+	// will fix them.
+	verboseLog.Info("statefulset status revisions", "updateRevision", sts.Status.UpdateRevision, "currentRevision", sts.Status.CurrentRevision)
+	var oldUnhealthyPods []*v1.Pod
+	for _, pod := range unhealthyPods {
+		if pod.Labels[appsv1.ControllerRevisionHashLabelKey] != sts.Status.UpdateRevision {
+			oldUnhealthyPods = append(oldUnhealthyPods, pod)
+		}
+	}
+
+	if len(oldUnhealthyPods) == 0 {
+		verboseLog.Info(" no statefulset unhealthy pods from old revisions")
+		return nil
+	}
+	verboseLog.Info("statefulset unhealthy pods from old revisions", "oldUnhealthyPods", listNames(oldUnhealthyPods))
+	log.Info("unhealthy pods from an old revision found, deleting them to unblock potential StatefulSet rollout")
```

Comment:
What's the difference between `verboseLog` and `log` and when would you use one vs the other?

Reply:
The log level. `logr` does not have a debug log level, only increased verbosity levels. `verboseLog.Info()` will not be shown to the user unless they enabled debug logs.